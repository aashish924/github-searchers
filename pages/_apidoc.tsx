import React from 'react';

import SwaggerUI from 'swagger-ui-react';
import 'swagger-ui-react/swagger-ui.css';

const apiDoc = () => {
    return <SwaggerUI url="/swagger.json" />;
};

export default apiDoc;
