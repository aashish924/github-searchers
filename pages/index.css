html,
body,
body > div:first-child,
div#__next,
div#__next > div,
div#__next > div > div,
div#__next > div > div > div {
  width: 100%;
  height: 100%;
}