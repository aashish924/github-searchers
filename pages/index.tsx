import React, { useState, useEffect }  from 'react';
import { connect, ConnectedProps } from "react-redux";
import styled, {CSSProperties} from 'styled-components';

import Card from "../components/Card";
import FlexGrid from '../components/FlexGrid';
import User from '../components/User';
import Repository from '../components/Repository';
import Layout from '../components/Layout';
import Search from '../components/Search';

import './index.css'


const AppHeader = styled.header`
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 60px;
  margin: 4px;
`;

const Brand = styled.div`
  display: flex;
  flex-direction: column;
  align-self: center;
  width: 100%;
`;

const AppName = styled.h1`
  font-weight: 600;
  font-size: 20px;
  margin: 0px;
`;

const Hint = styled.h3`
  color: gray;
  margin: 0px;
  font-size: 16px;
  font-weight: 200;
`;

const Loader = styled.span`
  position: relative;
  display: inline-block;
  text-align: center;
  width: 120px;
  height: 24px;
  left: calc(30% - 120px);
  margin: 16px;
  overflow: hidden;
  line-height: 1.5em;
  font-size: 18px;
  font-weight: 400;
  
  &::after {
    display: inline-table;
    white-space: pre;
    content: "🕐\\A🕑\\A🕒\\A🕓\\A🕔\\A🕕\\A🕖\\A🕗\\A🕘\\A🕙\\A🕚\\A🕛";
    animation: spin12 3s steps(12) infinite;
    width: 1.5em;
  }
  
  @keyframes spin12 {
    to {
      transform: translateY(-18.0em);
    }
  }
`;


const mapStateToProps = (state) => ({
  search: state.search,
});

const connector = connect(mapStateToProps, null);
type PropsFromRedux = ConnectedProps<typeof connector>;


const SearchResult = (searchType, result) => {
  const CardContent =  searchType == 'user' ? User : Repository;

  return <FlexGrid>
    {result.items.map( (item, i) => <Card key={i}><CardContent data={item}/></Card>)}
  </FlexGrid>
};


const IndexPage: React.FC<PropsFromRedux> = (props) => {
  const [searchState, setSearchState] = useState("initial");

  useEffect(() => {
    if(props.search) {
      setSearchState(props.search?.state);
    }
  });

  const initialStateStyle: CSSProperties = {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    verticalAlign: 'middle'
  };

  return <Layout title="Github Searcher">
    <div style={searchState !== "loaded" ? initialStateStyle : { marginTop: "24px" }}>
      <div>
        <AppHeader>
          <img src="/images/github.svg" style={{padding: "0 8px", width: "40px", height: "40px", marginTop: "-8px" }}/>
          <Brand>
            <AppName>Gihub Searchers</AppName>
            <Hint>Search users or repositories below</Hint>
          </Brand>
        </AppHeader>

        <Search />

        {(searchState === "loaded" && props.search.searchResult) ? SearchResult(props.search.searchType, props.search.searchResult): "" }
        {searchState === "loading" ? <Loader>Loading...</Loader> : ""}
      </div>
    </div>

  </Layout>
};

export default connector(IndexPage);
