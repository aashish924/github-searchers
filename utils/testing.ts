import { NextApiRequest, NextApiResponse } from 'next';
import User from '../shared/user/user';
import Repository from '../shared/repository/repository';
import { SearchResult } from './backend/github';
import { cache } from './backend/api';


export const mockRequest = ({method, json, headers}: {method?: string, json?: any, headers?: any}) => {
  return {
    method: method || 'GET',
    body: json || {},
    headers: headers || {}
  } as NextApiRequest
};


export const mockResponse = () => {
  const res = { } as NextApiResponse;
  res.status = jest.fn().mockReturnValue(res);
  res.json = jest.fn().mockReturnValue(res);
  return res;
};


export const mockUser = ({name, avatarUrl, url}: {name?: string, avatarUrl?: string, url?: string}): User => {
  const user: User = {
    name: name || 'John Doe',
    url: url || 'https://github.com/benzid-wael/',
    avatarUrl: avatarUrl || 'https://avatars0.githubusercontent.com/u/4288931?s=60&v=4'
  };
  return user;
};


export const mockRepository = (): Repository => {
  return {
    name: 'Github Searchers',
    repositoryUrl: 'https://github.com/benzid-wael/github-searchers',
    author: {
      name: 'John Doe',
      url: 'https://github.com/benzid-wael/',
      avatarUrl: 'https://avatars0.githubusercontent.com/u/4288931?s=60&v=4'
    },
  } as Repository;
};


export const mockUserSearchResult = (): SearchResult<User> => {
  return {
    meta: {
      total: 1,
      prev: null,
      next: null,
      last: 1
    },
    items: [
        mockUser({})
    ]
  }
};


export const mockRepositorySearchResult = (): SearchResult<Repository> => {
  return {
    meta: {
      total: 1,
      prev: null,
      next: null,
      last: 1
    },
    items: [
        mockRepository()
    ]
  }
};


export class FakeJsonLogger {

  // @ts-ignore
  @cache({ prefix: 'test', version: 1 })
  log(message: string, error?: string) {
    if (error) {
      throw new Error(error);
    }

    return {message: message}
  }

}