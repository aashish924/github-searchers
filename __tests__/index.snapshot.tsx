import * as React from 'react';

import { IndexPage } from '../pages/index';
import { render } from '../utils/testing/render';


describe('IndexPage', () => {
    it.each`
        state
        ${'initial'}
        ${'loading'}
        ${'loaded'}
    `('should render on $state state', ({ state }) => {
        // given
        jest.mock('../components/Search', () => () => 'Search');
        const initialState = { search: { state: state } };
        // when
        const tree = render(<IndexPage />, { initialState: initialState }).toJSON();
        // then
        expect(tree).toMatchSnapshot(`indexpage-${state}`);
    })
})