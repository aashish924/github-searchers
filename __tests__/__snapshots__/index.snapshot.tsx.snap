// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexPage should render on initial state: indexpage-initial 1`] = `
<div>
  <div
    className="sc-AxhUy jZpCAT"
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
          "verticalAlign": "middle",
        }
      }
    >
      <div>
        <header
          className="sc-fznyAO knaQiQ"
        >
          <img
            src="/images/github.svg"
            style={
              Object {
                "height": "48px",
                "marginTop": "-4px",
                "padding": "0 4px",
                "width": "48px",
              }
            }
          />
          <div
            className="sc-fznKkj dXJZUZ"
          >
            <h1
              className="sc-fznZeY kkymqc"
            >
              Gihub Searchers
            </h1>
            <h3
              className="sc-fzokOt ieIhLB"
            >
              Search users or repositories below
            </h3>
          </div>
        </header>
        <div
          className="sc-fzoLsD gdXpxU"
        >
          <div
            className="sc-fzpans kXNqgt"
          >
            <input
              autoComplete="off"
              className="sc-fzozJi bTtMmK"
              name="text"
              onChange={[Function]}
              placeholder="Start typing to search ..."
            />
            <div
              className="custom-select-wrapper"
            >
              <div
                className="custom-select"
              >
                <div
                  className="custom-select__trigger"
                  onClick={[Function]}
                >
                  <span>
                    user
                  </span>
                  <div
                    className="arrow"
                  />
                </div>
                <div
                  className="custom-options"
                >
                  <span
                    className="custom-option"
                    data-value="user"
                    onClick={[Function]}
                  >
                    Users
                  </span>
                  <span
                    className="custom-option"
                    data-value="repository"
                    onClick={[Function]}
                  >
                    Repositories
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            style={
              Object {
                "color": "red",
                "margin": "4px",
                "textAlign": "center",
              }
            }
          />
        </div>
        
        
      </div>
    </div>
  </div>
</div>
`;

exports[`IndexPage should render on loaded state: indexpage-loaded 1`] = `
<div>
  <div
    className="sc-AxhUy jZpCAT"
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
          "verticalAlign": "middle",
        }
      }
    >
      <div>
        <header
          className="sc-fznyAO knaQiQ"
        >
          <img
            src="/images/github.svg"
            style={
              Object {
                "height": "48px",
                "marginTop": "-4px",
                "padding": "0 4px",
                "width": "48px",
              }
            }
          />
          <div
            className="sc-fznKkj dXJZUZ"
          >
            <h1
              className="sc-fznZeY kkymqc"
            >
              Gihub Searchers
            </h1>
            <h3
              className="sc-fzokOt ieIhLB"
            >
              Search users or repositories below
            </h3>
          </div>
        </header>
        <div
          className="sc-fzoLsD gdXpxU"
        >
          <div
            className="sc-fzpans kXNqgt"
          >
            <input
              autoComplete="off"
              className="sc-fzozJi bTtMmK"
              name="text"
              onChange={[Function]}
              placeholder="Start typing to search ..."
            />
            <div
              className="custom-select-wrapper"
            >
              <div
                className="custom-select"
              >
                <div
                  className="custom-select__trigger"
                  onClick={[Function]}
                >
                  <span>
                    user
                  </span>
                  <div
                    className="arrow"
                  />
                </div>
                <div
                  className="custom-options"
                >
                  <span
                    className="custom-option"
                    data-value="user"
                    onClick={[Function]}
                  >
                    Users
                  </span>
                  <span
                    className="custom-option"
                    data-value="repository"
                    onClick={[Function]}
                  >
                    Repositories
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            style={
              Object {
                "color": "red",
                "margin": "4px",
                "textAlign": "center",
              }
            }
          />
        </div>
        
        
      </div>
    </div>
  </div>
</div>
`;

exports[`IndexPage should render on loading state: indexpage-loading 1`] = `
<div>
  <div
    className="sc-AxhUy jZpCAT"
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
          "verticalAlign": "middle",
        }
      }
    >
      <div>
        <header
          className="sc-fznyAO knaQiQ"
        >
          <img
            src="/images/github.svg"
            style={
              Object {
                "height": "48px",
                "marginTop": "-4px",
                "padding": "0 4px",
                "width": "48px",
              }
            }
          />
          <div
            className="sc-fznKkj dXJZUZ"
          >
            <h1
              className="sc-fznZeY kkymqc"
            >
              Gihub Searchers
            </h1>
            <h3
              className="sc-fzokOt ieIhLB"
            >
              Search users or repositories below
            </h3>
          </div>
        </header>
        <div
          className="sc-fzoLsD gdXpxU"
        >
          <div
            className="sc-fzpans kXNqgt"
          >
            <input
              autoComplete="off"
              className="sc-fzozJi bTtMmK"
              name="text"
              onChange={[Function]}
              placeholder="Start typing to search ..."
            />
            <div
              className="custom-select-wrapper"
            >
              <div
                className="custom-select"
              >
                <div
                  className="custom-select__trigger"
                  onClick={[Function]}
                >
                  <span>
                    user
                  </span>
                  <div
                    className="arrow"
                  />
                </div>
                <div
                  className="custom-options"
                >
                  <span
                    className="custom-option"
                    data-value="user"
                    onClick={[Function]}
                  >
                    Users
                  </span>
                  <span
                    className="custom-option"
                    data-value="repository"
                    onClick={[Function]}
                  >
                    Repositories
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            style={
              Object {
                "color": "red",
                "margin": "4px",
                "textAlign": "center",
              }
            }
          />
        </div>
        
        
      </div>
    </div>
  </div>
</div>
`;
